internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,363 @@
+//
+// The consumer is expect to pipe these connections together.
+type connectionGetter struct {
+	logger Logger
+}
+
+// NewConnectionGetter returns a new connectionGetter.
+func NewConnectionGetter(l Logger) *connectionGetter {
+	return &connectionGetter{l}
+}
+
+// GetControllerConnection initiates an SSH connection to the target ctrlAddress.
+func (w *connectionGetter) GetControllerConnection(password, ctrlAddress string) (ssh.Channel, error) {
+	// TODO(ale8k): Watch will return host key in subsequent PR.
+	sshConfig := &ssh.ClientConfig{
+		User:            ControllerSSHUser,
+		Auth:            []ssh.AuthMethod{ssh.Password(password)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(), // TODO(ale8k): Fill in host key here
+	}
+
+	client, err := ssh.Dial("tcp", ctrlAddress, sshConfig)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// TODO(ale8k): Make this a constant that both the server and session worker can use.
+	ch, in, err := client.OpenChannel("juju-tunnel", nil)
+	if err != nil {
+		return nil, err
+	}
+	go ssh.DiscardRequests(in)
+
+	return ch, nil
+}
+
+// GetSSHConnection performs a stand TCP dial to the SSHD running on the machine.
+func (w *connectionGetter) GetSSHDConnection() (net.Conn, error) {
+	etcPath := "/etc/ssh/sshd_config"
+	openSSHPath := "/usr/share/openssh/sshd_config"
+	port := w.getLocalSSHPort(etcPath, openSSHPath)
+	u, err := url.Parse("localhost" + port)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	localSSHD, err := net.Dial("tcp", u.String())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return localSSHD, nil
+}
+
+// getLocalSSHPort parses the local sshd_config file to get the port sshd is listening on.
+// It will try all the provided paths.
+// If it cannot get it for any reason, it'll log the error and return the default port 22.
+func (w *connectionGetter) getLocalSSHPort(filePaths ...string) string {
+	port := "22"
+
+	for _, filePath := range filePaths {
+		file, err := os.Open(filePath)
+		if err != nil {
+			w.logger.Errorf("Error opening sshd_config file: %v", err)
+			return port
+		}
+		defer file.Close()
+
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := strings.TrimSpace(scanner.Text())
+
+			if strings.HasPrefix(line, "#") || line == "" {
+				continue
+			}
+
+			if strings.HasPrefix(line, "Port") {
+				fields := strings.Fields(line)
+				if len(fields) == 2 {
+					port = fields[1]
+				}
+				break
+			}
+		}
+
+		if err := scanner.Err(); err != nil {
+			w.logger.Errorf("Error reading sshd_config file: %v", err)
+			return port
+		}
+		return port
+	}
+
+	return port
+}
+
+// cancellableReadWriter provides a means to cancel a read or write operation via context.
+type cancellableReadWriter struct {
+	ctx context.Context
+	rw  io.ReadWriter
+}
+
+// newCancellableReadWriter returns a new cancellableReadWriter.
+func newCancellableReadWriter(ctx context.Context, rw io.ReadWriter) *cancellableReadWriter {
```

Comment:
I don't feel the need for constructors on non-exported types, but you can leave this if it's convenient.